torch/ao/quantization/pt2e/prepare.py

```diff
@@ -5,28 +5,223 @@
     _get_output_act_obs_or_fq,
     _get_dtype_and_is_dynamic,
     _insert_obs_or_fq,
-    _maybe_insert_output_observer_for_node,
     _save_state,
     _is_activation_post_process_node,
-    _get_qspec_for_arg,
+    _create_obs_or_fq_from_qspec,
 )
 from torch.fx import (
     GraphModule,
+    Graph,
     Node,
 )
 from torch.fx.node import Argument
 
 from torch.ao.quantization import QConfigMapping
 from torch.ao.quantization.qconfig import QConfigAny
 from torch.ao.quantization.fx.custom_config import PrepareCustomConfig
-from typing import Dict, Tuple, Union, Any
+from typing import Dict, Tuple, Union, Any, Optional
 from torch.ao.quantization.quantizer import (
     QuantizationAnnotation,
     EdgeOrNode,
     SharedQuantizationSpec,
+    QuantizationSpecBase,
 )
 from torch.ao.quantization import ObserverOrFakeQuantize
 
+# TODO: make pt2e folder private?
+__all__ = [
+    "prepare",
+]
+
+def _find_root(edge_or_node: EdgeOrNode, shared_with_map: Dict[EdgeOrNode, EdgeOrNode]) -> EdgeOrNode:
+    """Find the root node for the sharing tree
+    Args:
+        edge_or_node: edge/node that we want to find the root
+        shared_with_map: each edge/node points to the parent, the root node will points to itself
+
+    Returns:
+        root edge/node
+    """
+    parent = shared_with_map[edge_or_node]
+    if parent == edge_or_node:
+        return edge_or_node
+    root = _find_root(parent, shared_with_map)
+    # path compression
+    shared_with_map[edge_or_node] = root
+    return root
+
+def _union(a: EdgeOrNode, b: EdgeOrNode, shared_with_map: Dict[EdgeOrNode, EdgeOrNode]) -> None:
+    """Merge the subtree for `b` with `a`, the order is important here
+    """
+    root_a = _find_root(a, shared_with_map)
+    root_b = _find_root(b, shared_with_map)
+    # union the two trees
+    shared_with_map[root_b] = root_a
```

Comment:
can you rename the args a and b to `parent` and `child` because thats what you are doing here.